### DISCUSSION TITLE
Pagination + Filter across 2 or more Models: Denormalization or Views?

### DISCUSSION THREAD

**Discussion by rwieruch** [prisma/prisma#20290] [Q&A] [answered]
These are my simplified models.

```prisma
model Guest {
  id           String    @id @default(cuid())
  orgId        String
  authorId     String?
  createdAt    DateTime  @default(now())
  updatedAt    DateTime  @updatedAt
  bookings     Booking[]
  email        String
}

model Booking {
  id               String            @id @default(cuid())
  orgId            String
  authorId         String?
  createdAt        DateTime          @default(now())
  updatedAt        DateTime          @updatedAt
  guest            Guest             @relation(fields: [guestId], references: [id])
  guestId          String
  checkInDate      String
  checkOutDate     String

  @@index([guestId])
}
```

In a simplified version, the client requests paginated bookings in a table component which include the guest (columns: guestEmail, bookingCheckInDate, bookingCheckOutDate). From the table component, it is also possible to filter on the server for email, checkInDate and checkOutDate through Prisma. Now on the server, I ran into the issue that I have to paginate across two models. 

---

Since this is only a simplified depiction of the problem, and my models have more fields, and there are not only 2 models involved but 4, I came to the following solution: The Guest model's email field needs to get duplicated in the Booking's model and there is a mechanism in place to keep the emails always in sync. Now I can apply the pagination/filter only on the bookings. However, I don't have a single source of truth anymore and having denormalized data in the DB does not feel 100% correct.

---

So even though my solution works, would there be a more idiomatic approach to my problem or is denormalization already the solution to it? The last day I came across database views and saw that Prisma offers these as well. Could I use a view (e.g. BookingWithGuest view) instead of denormalization? And would the view allow me to still paginate + filter?

Thanks for helping me out here! :)

**Comment by ludralph** [comment, marked as answer]
Hi @rwieruch 👋 

Your approach of denormalizing the data by duplicating the `email` field in the `Booking` model is one way to solve the problem of paginating across two models. However, as you mentioned, it does introduce some data redundancy.

Using database `Views` can be a more idiomatic approach to solving this problem while maintaining data integrity. A database `View` is a virtual table that is based on the result of a query. In your case, you can create a [View](https://www.prisma.io/docs/concepts/components/prisma-schema/views) called `BookingWithGuest` that joins the `Booking` and `Guest` tables and includes the necessary columns for the table component. You must first create the `View` in the underlying database, either manually or as part of a migration using  `migrate dev --create-only`. Once the `View` is created, you can use it as a regular table in your Prisma queries. You can paginate and filter the results using the columns from the `View`, just like you would with any other table. 

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by rwieruch** [reply]
Thanks @ludralph . This helps! Can I ask two follow ups quickly here?

- Are views supported for MySQL/PlanetScale?
- The [documentation](https://www.prisma.io/docs/concepts/components/prisma-schema/views)  is not clear about how the `view UserInfo` id is the id from the User and not from the Profile. The view never defines where this id comes from.

**Reply by ludralph** [reply]
Hi @rwieruch, 

Sure, happy to answer your follow-up questions!

> Are views supported for MySQL/PlanetScale?

Yes, views are supported for MySQL. As for PlanetScale, many aspects of using Prisma with PlanetScale are just like using Prisma with any other relational database. However, the documentation does not explicitly mention support for views with PlanetScale.

> The [documentation](https://www.prisma.io/docs/concepts/components/prisma-schema/views) is not clear about how the view UserInfo id is the id from the User and not from the Profile. The view never defines where this id comes from.

In the view UserInfo example, the id field is assumed to be from the User model. This is because the view block in Prisma schema is a representation of a view in your database, and the view in your database would define where the id comes from. When you create a view in your database, you would specify the source of each field. For example, in SQL, you might create a view like this:
```sql
CREATE VIEW UserInfo AS
SELECT u.id, u.email, u.name, p.bio
FROM User u
JOIN Profile p ON u.id = p.userId;
```
In this SQL statement, u.id specifies that the `id` field in the UserInfo view comes from the `id` field in the `User` table. When you introspect this view with Prisma, it will generate a view block in your Prisma schema with an `id` field, representing the `id` field from the `User` table.

If this answers your question, it would be great if you could mark this Discussion as answered to indicate that it has been resolved.

Otherwise, please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by ludralph** [reply]
Hi @rwieruch 

Do you have any more questions around this issue? If no, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!
